stdlib/args.ysh

```diff
@@ -24,140 +24,159 @@
+#   proc namespace only within `Args`
+# - need :: because vars start with _
+# - We need a mechanism to share state between Args and `flag`/`arg` that is
+#   run in the block passed to it
+#   - It would be nice if _this was a magic param?
+#     - it would be nice if 'flag' was a "method" on something?
+#     - that's what it's doing, it's mutating "self"
+#     - _this is the thing we're currently creating
+
+proc Args (spec Ref;;; block) {
+  ## Create an args spec which can be passed to parseArgs.
+  ##
+  ## Example:
+  ##
+  ##   # NOTE: :spec will create a variable named spec
+  ##   Args :spec {
+  ##     flag -v --verbose ('bool')
+  ##   }
+  ##
+  ##   :: parseArgs(spec, ARGV)
+
+  setref spec = {
+    flags: [],
+    args: [],
+    rest: null,
   }
 
-  case (len(strs)) {
-    (1) {
-      echo
-    }
-    (2) {
-      echo
+  eval (block) | while read line {
+    json read parsed <<< $line
+    var spec_frag = dict(parsed)
+
+    case (spec_frag.tag) {
+      flag {
+        _ spec.flags->append(spec_frag.node)
+      }
+      arg {
+        _ spec.args->append(spec_frag.node)
+      }
+      rest {
+        setvar spec.rest = spec_frag.node
+      }
     }
   }
-
-  var result = {short, long, type, default}
-
-  # TODO:
-  # - need :: because vars start with _
-  # - It would be nice if _this was a magic param?
-  #   - it would be nice if 'flag' was a "method" on something?
-  #   - that's what it's doing, it's mutating "self"
-  #   - _this is the thing we're currently creating
-
-  _ _this.flags->append(result)
-
-  var this = getvar('_this', scope='dynamic')
-  _ this.flags->append(result)
 }
 
-proc __arg (name, desc ; type ; default Any = null) {
-  # Type is always string?  Or support conversion?
-  # I guess there's no reason not to
-  #
-  # Int or Float
-
-  echo 'arg'
+proc flag (short, long; type='bool', default=null) {
+  ## Declare a flag within an `Args`.
+  ##
+  ## Examples:
+  ##
+  ##   Args :spec {
+  ##     flag -v --verbose
+  ##     flag -n --count ('int', default=1)
+  ##     flag -f --file ('str')
+  ##   }
+
+  # TODO: help, make type take in a type object (ie. Bool, Int, Str)
+  json write --pretty=F ({ tag: 'flag', node: { short, long, type, default, help: null } })
 }
 
-# TODO: create a builtin?
-func getProc(x) {
-  return (x)
+proc arg (name) {
+  ## Declare a positional argument within an `Args`.
+  ##
+  ## Examples:
+  ##
+  ##   Args :spec {
+  ##     arg file
+  ##   }
+
+  # TODO: type, default and help
+  json write --pretty=F ({ tag: 'arg', node: { name, type: 'str', default: null, help: null } })
 }
-# TODO: builtin?
-# Is this different than shvar?  How?
-# Is it a global or a local?
 
-# Maybe we can rename shvar IFS=: (foo={}) { }
-#
-# - shvar does LocalOnly
-# - but 'proc arg' will push a new stack frame?
-#   - so we need something like setref?
-#
-
-proc bind-vars (; ; ...bindings; block) {
-  echo
+proc rest (name) {
+  ## Take the remaining positional arguments within an `Args`.
```

Comment:
I wonder if we can just use python style docstrings, like 

`''' foo bar '''`

We didn't have those when I added the doc comments, which are `###`